lib/HLSL/HLOperationLower.cpp

```diff
@@ -7602,6 +7602,20 @@ void TranslateStructBufSubscriptUser(
         handle, ResKind, bufIdx, baseOffset, status, OP, DL);
     }
     BCI->eraseFromParent();
+  } else if (PHINode* Phi = dyn_cast<PHINode>(user)) {
```

Comment:
I'm a bit afraid this might not be the only place impacted by phi on handle leftover from lcssa.  I wonder if it wouldn't be better to explicitly clean these up with a pass before they will pose a problem instead?

Since these get introduced by lcssa dependency of loop-rotate, loop-unroll, dxil-loop-unroll (I think, and maybe others), should we clean these up explicitly after these passes, when lcssa form is no longer necessary?  Perhaps we could clean up all single predecessor phis (not just handles)?

Reply:
Is there a pre-dxilgen pass where we could clean up these phis?

Reply:
There's no pass that I know of specifically for this; normally they would be cleaned up by simplify-inst, but we don't want to run that on `/Od`.

Is there any harm to debuggability if we make a new pass to remove all single-pred (or maybe even single-incoming-value) phis?  This pass doesn't have to be HL or DXIL specific.

I also wonder if it's possible to have multi-incoming-value phis that would break this (or other) code, which we simply haven't hit because we don't generate the code that way normally (due to HLSL language limitations mostly).  Code like this seems fragile, considering language features we may want to add.

Reply:
> Should we do a separate pass?

I think I'd prefer a pass, but second (third?) opinions are welcome.

The way I see it, the phi here was breaking an assumption which could likely be made elsewhere as well.  Without /Od, we are relying on simplify-inst to clean these up.  A pass to clean these up when we don't want full simplify-inst would help keep that assumption from being broken unless it really is the case that requires much more work to fix (multiple incoming values).  A targeted fix avoids the assumption for this special case in this code, but wouldn't help other places that make the same assumption.

One advantage of the pass is being able to make the pass handle slightly more complex cases that can still be simplified this way, such as a chain of single-value phis, or phis with multiple preds using the same incoming value, and recursive simplification of each.  Maybe we can even use it to sink GEPs or handle create/annotate calls in order to clean up multi-value phis, preserving downstream assumptions for these cases as well.

Reply:
Is there really a broad assumption that phi's won't occur? I would have expected more bugs around this area if that were the case.

I would prefer to make a smaller fix first to fix the issue we have in hand. That will be less risk overall. If we want to create a new normalizing pass we could do that as a follow up, but I'm not convinced this is a widespread issue.

Reply:
Discussed this offline some more. We already have the `dxil-o0-simplify-inst` pass as a replacement for simplifyinst when compiling with O0. We can run that pass in place of simplifyinst before dxilgen to fix this bug.

Updated the PR with that new strategy.

Reply:
The new version is failing in some internal tests because it generates invalid dxil (a phi with an undef input). It seems like the DxilValueCache is maybe getting populated with more information that it was not previously, and a later pass (dxil-remove-dead-blocks) is taking advantage of it to simplify the code. 

```
error: Instructions should not read uninitialized value.
note: at '%472 = phi float [ fdiv (float undef, float undef), %469 ], [ %432, %470 ]' in block '#10'
```

Have not had time to debug to see what exactly is going wrong.

Reply:
I'm reverting this PR back to the previous targeted fix strategy. I did not have time to investigate the failures from running the `dxil-o0-simplify-inst` more widely. Going to get this fix in since it at least leaves us in a better position than before.